plugins-builtin/src/main/scala/za/co/absa/enceladus/plugins/builtin/errorinfo/mq/ErrorInfoSenderPlugin.scala

```diff
@@ -0,0 +1,178 @@
+ * limitations under the License.
+ */
+
+package za.co.absa.enceladus.plugins.builtin.errorinfo.mq
+
+import org.apache.log4j.LogManager
+import org.apache.spark.sql.functions.{col, explode, lit, size, struct}
+import org.apache.spark.sql.types.DataTypes
+import org.apache.spark.sql.{DataFrame, Encoders}
+import za.co.absa.enceladus.plugins.api.postprocessor.PostProcessor
+import za.co.absa.enceladus.plugins.builtin.common.mq.kafka.KafkaConnectionParams
+import za.co.absa.enceladus.plugins.builtin.errorinfo.DceErrorInfo
+import za.co.absa.enceladus.plugins.builtin.errorinfo.mq.ErrorInfoSenderPlugin.SingleErrorStardardized
+import za.co.absa.enceladus.plugins.builtin.errorinfo.mq.kafka.KafkaErrorInfoPlugin
+import za.co.absa.enceladus.utils.schema.SchemaUtils
+import ErrorInfoSenderPlugin._
+import za.co.absa.enceladus.plugins.builtin.errorinfo.params.ErrorInfoPluginParams
+import za.co.absa.enceladus.plugins.builtin.errorinfo.params.ErrorInfoPluginParams.ErrorSourceId
+import za.co.absa.enceladus.utils.error.ErrorMessage.ErrorCodes
+
+import scala.util.{Failure, Success, Try}
+
+
+case class ErrorInfoSenderPlugin(connectionParams: KafkaConnectionParams,
+                                 keySchemaRegistryConfig: Map[String, String],
+                                 valueSchemaRegistryConfig: Map[String, String]) extends PostProcessor {
+
+  private val log = LogManager.getLogger(classOf[ErrorInfoSenderPlugin])
+
+  override def close(): Unit = {}
+
+  /**
+   * When data is ready, the error info record(s) are pusblished to kafka.
+   *
+   * @param dataFrame error data only.
+   * @param params    Additional key/value parameters provided by Enceladus.
+   */
+  override def onDataReady(dataFrame: DataFrame, params: Map[String, String]): Unit = {
+    if (!SchemaUtils.fieldExists(ColumnNames.enceladusRecordId, dataFrame.schema)) {
+      throw new IllegalStateException(
+        s"${this.getClass.getName} requires ${ColumnNames.enceladusRecordId} column to be present in the dataframe!"
+      )
+    }
+
+    val errorInfoParams = Try(ErrorInfoPluginParams.fromMap(params)) match {
+      case Success(params) => params
+      case Failure(e) =>
+        throw new IllegalArgumentException(s"Incompatible parameter map supplied for ${ErrorInfoSenderPlugin.getClass.getName}: $params", e)
+    }
+
+    val dfWithErrors = getIndividualErrors(dataFrame, errorInfoParams)
+    val forKafkaDf = prepareDataForKafka(dfWithErrors, errorInfoParams)
+    sendErrorsToKafka(forKafkaDf)
+  }
+
+  /**
+   * Processes the `dataFrame` - errors are exploded (one error = one line) and filtered to conform to the error source (standardization/conformance)
+   * @param dataFrame standardization/conformance output
+   * @param params plugin processing parameters
+   * @return DF with exploded errors and corresponding to the given error source
+   */
+  def getIndividualErrors(dataFrame: DataFrame, params: ErrorInfoPluginParams): DataFrame = {
```

Comment:
I'm very curious about this implementation. Not sure that I understand all of it.

How about a quick walkthrough? Who wants to join?

Reply:
Well, it does number of things:
 - input projection for `endceladus_record_id`, `reportDate`, and `errCol`.
 - explodes the `errCol` to have only 1 error per line
 - selection over the `errCol`'s errorCode - we maintain a 1-many mapping between source (Std/Conf) and errorCodes - only the corresponding errors are filtered
 - mapping the dataframe entries (with the help of the plugin param values) to `DceError`s.

(The typical usage is that `DceError`s are then directly convertible to agreed-upon confluent_avro format and sent to Kafka).

If this description is not clear, we can discuss/explain that further.

Are you pointing to the fact that some of the functionality could be generalized and perhaps used elsewhere?

Reply:
Thanks for the explanation! The code is perfectly good and I've learned something from it. And it is very clear as well.

I just have a couple of questions. More for self-learning. Maybe others are interested as well.

I'll ask one particular question in the comments.